File: src/03_Functions_And_Scope/PassByRef_Vs_Value.java
/**
 * Level 3: Functions & Call Stack
 * File: PassByRef_Vs_Value.java
 * Topic: Pass-by-value semantics in Java (Primitives vs Object Reference copies)
 */

public class PassByRef_Vs_Value {
    static class Box { int val; Box(int v){ this.val = v; } }

    // Primitives are passed by value (Copy made)
    static void tryModifyPrimitive(int num) {
        num = 999;
    }

    // Reference pointer copy is passed (Modifies heap object state)
    static void modifyBoxContent(Box b) {
        b.val = 999; // Mutates heap object
    }

    // Reference reassignment does NOT affect caller
    static void reassignBox(Box b) {
        b = new Box(777); // Local reference reassigned only!
    }

    public static void main(String[] args) {
        int x = 5;
        tryModifyPrimitive(x);
        System.out.println("Primitive after method: " + x); // 5

        Box b = new Box(5);
        modifyBoxContent(b);
        System.out.println("Box val after modify: " + b.val); // 999

        reassignBox(b);
        System.out.println("Box val after reassign: " + b.val); // Still 999!
    }
}

File: src/03_Functions_And_Scope/PassByRef_Vs_Value.cpp
/**
 * Level 3: Functions & Call Stack
 * File: PassByRef_Vs_Value.cpp
 * Topic: Pass-by-value vs pass-by-reference (&) vs pass-by-pointer (*)
 */

#include <iostream>

// 1. Pass by Value (Copy created on stack frame)
void incrementVal(int a) {
    a += 10;
}

// 2. Pass by Reference (Modifies original variable)
void incrementRef(int &a) {
    a += 10;
}

// 3. Pass by Pointer (Uses memory address)
void incrementPtr(int *a) {
    if (a != nullptr) {
        *a += 10;
    }
}

int main() {
    int x = 5;
    
    incrementVal(x);
    std::cout << "After Pass-by-Value: " << x << " (Unchanged)\n"; // 5

    incrementRef(x);
    std::cout << "After Pass-by-Ref:   " << x << " (Updated!)\n";  // 15

    incrementPtr(&x);
    std::cout << "After Pass-by-Ptr:   " << x << " (Updated!)\n";  // 25

    return 0;
}
